### ISSUE TITLE
Performance issues with v0.13 vs. BVLC/master

### ISSUE THREAD

**Issue by lukeyeager** [NVIDIA/caffe#55]
*From https://github.com/NVIDIA/DIGITS/issues/389*

>My system:
1. Cuda 7.5
2. CudNN V3
3. 2x Titan X's

> I am transitioning from Caffe master from BVLC to Digits mainly for the ease of use. 
1. When I train the reference_caffenet model with Imagenet 1000 using Caffe master (1 Titan X), training 20 iterations take about 6sec to 6.5 seconds. 
2. I used the web installer of digits (which I believe comes with Cuda 7.0 and cudnnv3). If I start training on the same reference_caffenet model with Digits (1 Titan X), training 20 iterations takes about double the time (13 to 14 seconds). I also notice a lot of log messages like the one below:

> `18:29:17.241222 11321 blocking_queue.cpp:50] Data layer prefetch queue empty`

> I tried recompiling NVCaffe with `USE_CNMEM := 0` and `USE_CNMEM := 1` (as that was the only difference between NVidia's Caffe and BVLC's Caffe makefiles) but there was no difference in performance. 

> Any tips on how to debug this? I really want to be able to use Digits with 2 GPUs. I don't see the above problem  with the Caffe master from BVLC. 

> Siddharth
@siddharthm83

**Comment by lukeyeager** [comment]
Hi @siddharthm83,

That's weird, I wouldn't expect there to be a big difference in performance between v0.13 and BVLC/master. Are you sure you're calculating the number of iterations correctly? I only ask because DIGITS talks about "epochs" instead of "iterations".

Can you try out the `caffe-0.14` branch for me? That includes a much more recent sync from BVLC/master. Note that you'll need a recent version of DIGITS to work with that version of NVcaffe (you need this pull request - https://github.com/NVIDIA/DIGITS/pull/336).

**Comment by thatguymike** [comment]
"Data layer prefetch queue empty" implies a that the data reading itself is not keeping up with training.  Is the DIGITS database location on the same IO device as your manual tests?

**Comment by siddharthm83** [comment]
@lukeyeager , I tired `caffe-0.14` branch. I don't see the  `Data layer prefetch queue empty` seems to have dissappeared. But for some reason the performance is worse. The time for training 20 iterations is now close to 20 seconds (from earlier of 14secs). 
I am training for the same number of epochs (90) with the same batch size of 256 for all cases. 
'I1029 14:00:52.977888 20055 solver.cpp:243] Iteration 40, loss = 6.97762
I1029 14:00:52.977973 20055 solver.cpp:259]     Train net output #0: loss = 6.97762 (* 1 = 6.97762 loss)
I1029 14:00:52.977982 20055 solver.cpp:590] Iteration 40, lr = 0.01
I1029 14:01:10.093080 20055 solver.cpp:243] Iteration 60, loss = 6.93633
I1029 14:01:10.093106 20055 solver.cpp:259]     Train net output #0: loss = 6.93633 (* 1 = 6.93633 loss)
I1029 14:01:10.093112 20055 solver.cpp:590] Iteration 60, lr = 0.01`

Below, you can see training time for the same model  but with BVLC/master
`I1029 15:17:08.761509 20493 solver.cpp:236] Iteration 40, loss = 6.93335
I1029 15:17:08.761553 20493 solver.cpp:252]     Train net output #0: loss = 6.93335 (* 1 = 6.93335 loss)
I1029 15:17:08.761564 20493 sgd_solver.cpp:106] Iteration 40, lr = 0.01
I1029 15:17:14.897974 20493 solver.cpp:236] Iteration 60, loss = 6.95387
I1029 15:17:14.898021 20493 solver.cpp:252]     Train net output #0: loss = 6.95387 (* 1 = 6.95387 loss)
I1029 15:17:14.898032 20493 sgd_solver.cpp:106] Iteration 60, lr = 0.01`

The only difference seems to be the databases. For BVLC/master, I created the database from imagenet1000 dataset using caffe's scripts. For Digits, I created the database from the same images using the Digits GUI. I do notice that the database size is different (not sure why). The BVLC/master training database is about twice as large. Could this be the problem? Digits only seems to let me choose the dataset imported through digits. 

@thatmikeguy, I have a 4TB HDD that I am doing all my work on. So the IO device are the same.

**Comment by siddharthm83** [comment]
Ok, I think it was the database that was the problem. I tried to run BVLC/master using the database created by digits and I get 14 seconds for every 20 iteration which was what DIGITS gave me as well before I moved to caffe-0.14. 
`I1029 15:46:19.303618 20826 solver.cpp:236] Iteration 40, loss = 6.94188
I1029 15:46:19.303683 20826 solver.cpp:252]     Train net output #0: loss = 6.94188 (* 1 = 6.94188 loss)
I1029 15:46:19.303704 20826 sgd_solver.cpp:106] Iteration 40, lr = 0.01
I1029 15:46:33.006464 20826 solver.cpp:236] Iteration 60, loss = 6.9368
I1029 15:46:33.006608 20826 solver.cpp:252]     Train net output #0: loss = 6.9368 (* 1 = 6.9368 loss)
I1029 15:46:33.006628 20826 sgd_solver.cpp:106] Iteration 60, lr = 0.01
I1029 15:46:46.737506 20826 solver.cpp:236] Iteration 80, loss = 6.93274
I1029 15:46:46.737567 20826 solver.cpp:252]     Train net output #0: loss = 6.93274 (* 1 = 6.93274 loss)`

GPU utilization is not consistently 100% as well.

**Comment by lukeyeager** [comment]
Looks like I may have been a little too quick to move this to the NVcaffe module ...

Try creating your LMDB without encoding them as PNG images:

![lmdb-no-encoding](https://cloud.githubusercontent.com/assets/687269/10834559/8be21a06-7e57-11e5-989d-96598c6dc18c.jpg)

**Comment by lukeyeager** [comment]
I'll assume that switching to non-PNG data solved your problem. Please let me know otherwise.

**Comment by siddharthm83** [comment]
Yes, it did. Thank you!

**Comment by deliafc** [comment]
Hello, 
I'm getting the same message "blocking_queue.cpp:50] Data layer prefetch queue empty" and the fine tuning is extremely slow, but I'm not using LMBD, I use a .txt with the image paths. I checked the image sizes and the coding type and they are around 100K and jpg.
Any clue of what may be going wrong?
Thanks

**Comment by lukeyeager** [comment]
@deliafc which version of Caffe are you using?

Those "Data layer prefetch queue empty" messages don't necessarily mean that anything is wrong, and training could be slow for many reasons.

**Comment by deliafc** [comment]
@lukeyeager it's the version from october 2015

**Comment by lukeyeager** [comment]
You might try a newer release. [v0.14.5](https://github.com/NVIDIA/caffe/tree/v0.14.5) is our latest packaged release.

**Comment by westeast** [comment]
I slove the problem by changing the batches number to 16 from 32
